### DISCUSSION TITLE
Userdefined parameter functions which use pybamm symbols

### DISCUSSION THREAD

**Discussion by CaXYZ** [pybamm-team/PyBaMM#1870] [Q&A] [answered]
Hi all,

I have a question concerning a used-defined function, e.g., for OCP, to which I would like to pass additional arguments besides the pybamm symbol sto. So my question is: how can I realize this. 

For example, if my OCP-function is of the form:
my_ocp(sto,param1,param2,param3) 

My approach was to pass the function as a parameter:
...
param = pybamm.ParameterValues(chemistry=chemistry)
param['Positive electrode OCP [V]']=my_ocp(sto,param1,param2,param3) 
...

But I do not know to define sto in the right way, i.e., how to handle/get access to the symbol and implement this in my function.
In general, is there a way to get to know the symbols defined in pybamm and their meaning?

It would be very nice if someone can provide a hint how to manage this issue.
 
With best regards
Carsten

**Comment by rtimms** [comment]
What are the parameters it depends on? If they can be defined ahead of time (e.g. you know their values and want to run simulations with different values) you could wrap the function like

```python3
param1, param2, param3 = ...

def pos_ocp(sto):
    return my_ocp(sto, param1, param2, param3)


param = pybamm.ParameterValues(chemistry=chemistry)
param['Positive electrode OCP [V]']=pos_ocp(sto,param1,param2,param3)
```

You could make `param1` etc. `InputParameter` objects then you can set their values when you call `solve`. See [here](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/parameterization/parameter-values.ipynb) for an example.

**Reply by CaXYZ** [reply]
Hi @rtimms,
thank you for support and the link to the example. The parameters are known before a single simulation. These are (time-independent) fitting constants, which I would like to vary by means of an optimization algorithms. So the problem is in general about parameter optimization.

I now tested several ways to implement the solution suggested, but already failed to set up a minimum example (even without the additional parameters implemented):

```
import pybamm
import numpy as np
import matplotlib.pylab as plt
  

def nmc_ocp_Xu2019_modified_2(sto):
    # Values from Mohtat2020, might be more accurate
    u_eq = (
        4.3452
        - 1.6518 * sto
        + 1.6225 * (sto ** 2)
        - 2.0843 * (sto ** 3)
        + 3.5146 * (sto ** 4)
        - 2.2166 * (sto ** 5)
        - 0.5623e-4 * pybamm.exp(109.451 * sto - 100.006)
    )


    return u_eq


#Call half cell model
model_options={'working electrode':'positive'}
model=pybamm.lithium_ion.DFN(options=model_options)

# Defining chemistry
chemistry = pybamm.parameter_sets.Xu2019

param = pybamm.ParameterValues(chemistry=chemistry)

#sto=pybamm.Variable['Positive electrode SOC']
sto=model.variables['Positive electrode SOC']

#param.update({"testparam1": "[input]"},check_already_exists=False)
param['Positive electrode OCP [V]']=nmc_ocp_Xu2019_modified_2(sto)


simulation = pybamm.Simulation(model, parameter_values=param)
    
solutions=simulation.solve(t_eval=[0,1000])

simulation.plot()
```

So my problem is that I  do not know how to implement sto in the right way.

With best regards
Carsten

**Comment by rtimms** [comment, marked as answer]
This example might help 

```python3
import pybamm
from pybamm import exp, constants

# pick parameters
param = pybamm.ParameterValues("Ecker2015")

# define custom diffusivity function with some input parameters we can specify when solving
a = pybamm.InputParameter("a")
b = pybamm.InputParameter("b")
c = pybamm.InputParameter("c")


def D_n(sto, T):
    D_ref = a * exp(-b * sto) + c
    E_D_s = 3.03e4
    arrhenius = exp(-E_D_s / (constants.R * T)) * exp(E_D_s / (constants.R * 296))
    return D_ref * arrhenius

# update parameters
param["Negative electrode diffusivity [m2.s-1]"] = D_n

# set up simulation 
model = pybamm.lithium_ion.SPMe()
sim = pybamm.Simulation(model, parameter_values=param)

# solve, specifying inputs (you could wrap this call to solve in an optimizer)
sim.solve([0, 3600], inputs={"a":4.4e-13, "b":10.3, "c":8.2e-15})

# plot
sim.plot()
```

**Reply by CaXYZ** [reply]
Hi @rtimms,

this example was exactly what I was look for. Thanks a lot!

Best regards
